### PR TITLE
eth_sign toggle Update in advanced settings

### LINKED ISSUE #18586: Add friction to enable eth_sign
## Description
Following [#17308](https://github.com/MetaMask/metamask-extension/pull/17308) `eth_sign` has been disabled by default, but users can enable the confirmation method in Settings > Advanced. Enabling `eth_sign` puts the user at risk of [phishing scams](https://medium.com/chainlight/si-vis-pacem-para-bellum-exploring-metamask-phishing-4605425d80a7) that can result in loss of funds.

### Current
- When `eth_sign` is enabled and the user received an `eth_sign` signature request they are shown a _warning_ on the confirmation screen
- On the settings screen, the user is _informed_ that `eth_sign` makes for a phishing risk as it is an open-ended signing method. And to only sign this type of confirmation request if they can read what they are signing and trust the source
- When directed by a scammer to enable `eth_sign`, there are few obstacles or alarm bells for the user to do so (information only)

### Expected
- For both popup and fullscreens:
    - Changes to the Settings > Advanced screen
         - `eth_sign` toggle includes the copy "Recommended" when it's disabled and "not recommended" when its enabled. 
    - Addition of a modal with checkbox
         - Double opt in DS component
    - Addition of a a modal with a text input. 

##### 1. Open double opt-in modal with checkbox
- Given the user is on Settings > Advanced 
- And eth_sign is disabled
- When the user attempts to enable eth_sign
- Then the user is shown a modal with a double opt-in (checkbox + Enable button)

##### 2. Initial state of modal buttons
- Given the user selects eth_sign
- And a modal is presented
- When the checkbox in the modal is unchecked
- Then the `Enable` eth_sign button on this modal is inactive AND a `Cancel` button is active

##### 3. Check checkbox
- Given the user selects Enable eth_sign 
- And a modal is presented
- When the checkbox is checked
- Then the `Enable` button is active AND a `Cancel` button is active

##### 4. Enable eth_sign
- Given the checkbox on the double opt-in modal is checked
- When the user selects `Enable`
- Then the triple opt-in modal is shown

##### 5. Enter text to confirm
- Given the triple opt-in modal is shown, and text input field 
- When the user enters "I only sign what I understand"
- Then eth_sign is enabled

##### 6. Incorrect text
- Given the triple opt-in modal is shown, and text input field 
- When the user enters any text not matching "I only sign what I understand"
- The alert "This text is incorrect. Try again." is shown [go back to #5]

##### 6. Cancel and close modal
- Given the triple opt-in modal is shown
- When the user selects `Cancel`
- Then the modal is closed AND the toggle remains disabled

[Figma popup](https://www.figma.com/file/yb1yRlmbgeGhNiTJvapBTm/settings-eth_sign?node-id=0%3A1&t=cdMpoVnEIAtRj4gq-1)
[Figma Expanded screen](https://www.figma.com/file/yb1yRlmbgeGhNiTJvapBTm/settings-eth_sign?node-id=478-23682&t=uNjQS6kjGs2GE2fl-0)
[Prototype](https://www.figma.com/proto/yb1yRlmbgeGhNiTJvapBTm/settings-eth_sign?page-id=0%3A1&node-id=13-17610&viewport=-166%2C142%2C0.25&scaling=min-zoom&starting-point-node-id=13%3A17610)

### PR DESCRIPTION
**Pull request by NidhiKJha**
Currently,  eth_sign has been disabled by default, but users can enable the confirmation method in Settings > Advanced. Enabling eth_sign puts the user at risk of [phishing scams](https://medium.com/chainlight/si-vis-pacem-para-bellum-exploring-metamask-phishing-4605425d80a7) that can result in loss of funds. This PR is to add extra friction to enable eth_sign by adding Modal and text input. This PR also includes the analytics update for eth_sign

1. When user completes the flow to enable eth_sign toggle, we should fire the Settings Updated event with a enable_eth_sign boolean property with true as the value.

2. When user completes the flow to disable eth_sign toggle, we should fire the Settings Updated event with a enable_eth_sign boolean property with false as the value.

Fixes #18586 

### Before

https://user-images.githubusercontent.com/39872794/234841744-2096c3fb-ffaa-424a-ba5f-d424fae584e0.mov

### After

https://user-images.githubusercontent.com/39872794/235162191-2ff933b7-503d-433e-b36d-1e076d405401.mov

## Manual Testing Steps

-  Go to the Settings > Advanced screen
- Toggle on the eth_sign toggle button
- Complete the steps in the modal
- At any step if you cancel the flow or click outside of the modal, it won't toggle on the modal unless you click the enable button in the last step of modal
- To turn off the eth_sign just click the toggle button

## Pre-merge author checklist

- [x] I've clearly explained:
  - [x] What problem this PR is solving
  - [x] How this problem was solved
  - [x] How reviewers can test my changes
- [x] Sufficient automated test coverage has been added

## Pre-merge reviewer checklist

- [ ] Manual testing (e.g. pull and build branch, run in browser, test code being changed)
- [ ] PR is linked to the appropriate GitHub issue
- [ ] **IF** this PR fixes a bug in the release milestone, add this PR to the release milestone

If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.

### CODE REVIEW

**Comment by SaraCheikh** [comment]
@coreyjanssen for the last step before enabling the setting, are we okay removing "try again" from the error state copy since it will be shown while they are typing the text?

**Review by pedronfigueiredo — DISMISSED** [review]

**Comment by SaraCheikh** [comment]
This looks very clean @NidhiKJha ! small details for the UI: 

-The spacing between title and body text looks very tight. In design, the default spacing in the modal component is 24px.

 Screenshot 2023-04-28 at 13 56 27
 
 -We are using a secondary button (outlined) instead of a primary one (filled) for " cancel",  we are also using a size lg for these buttons which looks slightly bigger. 
 
Screenshot 2023-04-28 at 14 00 40

- @coreyjanssen has used sentence case in design for `(Not recommended)` and `(Recommended) ` int he toggle state. 

Screenshot 2023-04-28 at 14 08 26

**Comment by NidhiKJha** [comment]
> This looks very clean @NidhiKJha ! small details for the UI:
> 
> -The spacing between title and body text looks very tight. In design, the default spacing in the modal component is 24px.
> 
@SaraCheikh I have updated the changes. I have updated the screencast video in description and you can check the Modal [here ](https://output.circle-artifacts.com/output/job/7ed58095-4d11-4489-99bb-cedfc245e929/artifacts/0/storybook/index.html?path=/story/components-app-modals-ethsignmodal--default-story)

**Comment by coreyjanssen** [comment]
> @coreyjanssen for the last step before enabling the setting, are we okay removing "try again" from the error state copy since it will be shown while they are typing the text?

Works for me! Remove the period as well as "Try again" so it just reads `The text is incorrect` :)

**Review by darkwing — DISMISSED** [review]
I'm seeing the following error when I click to sign:

SCR-20230428-jnnt

Nothing happens upon first click. I have to click multiple times to actually sign.

**Comment by SaraCheikh** [comment]
> > This looks very clean @NidhiKJha ! small details for the UI:
> > -The spacing between title and body text looks very tight. In design, the default spacing in the modal component is 24px.
> 
> @SaraCheikh I have updated the changes. I have updated the screencast video in description and you can check the Modal [here ](https://output.circle-artifacts.com/output/job/7ed58095-4d11-4489-99bb-cedfc245e929/artifacts/0/storybook/index.html?path=/story/components-app-modals-ethsignmodal--default-story)

Buttons look great now @NidhiKJha! only two more details:
-Remove "try again" in the copy of the error state.
-Make spacing between title and body text 24px. 
Screenshot 2023-05-01 at 12 48 29

**Review by brad-decker — DISMISSED** [review]
So, the components/app/modals pattern and using dispatch to open modals should be deprecated and marked thusly. we do not want to extend this folder or its usages, at least the last time I checked. We had a popover component that can be used instead that doesn't require a dispatch to control showing it... unless that has changed recently? @georgewrmarshall what is the current approach recommended for modal implementation?

**Comment by NidhiKJha** [comment]
> I'm seeing the following error when I click to sign:

Hmm! I looked into this issue. This is not related to this PR but the sign button is not working. I have created a [separate issue here](https://github.com/MetaMask/metamask-extension/issues/18942)

**Comment by NidhiKJha** [comment]
> So, the components/app/modals pattern and using dispatch to open modals should be deprecated and marked thusly. we do not want to extend this folder or its usages, at least the last time I checked. We had a popover component that can be used instead that doesn't require a dispatch to control showing it... unless that has changed recently? @georgewrmarshall what is the current approach recommended for modal implementation?

Thanks for pointing it out Brad. I have used the Modal component since we have used this component for all the modals on the settings page and it's all handled in one file. I know that @georgewrmarshall is working on the new Modal component that will deprecate the old Modals once it's merged but I wasn't aware that we are deprecating this Modals with Popover.

File: ui/components/app/modals/eth-sign-modal/eth-sign-modal.js [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+  const trackEvent = useContext(MetaMetricsContext);
+  const handleCancel = () => {
+    hideModal();
+  };
+
+  const handleSubmit = () => {
+    dispatch(
+      setDisabledRpcMethodPreference(
+        'eth_sign',
+        !disabledRpcMethodPreferences?.eth_sign,
```
[the remark this answers is not in the document]
Comment (NidhiKJha) [reply ui/components/app/modals/eth-sign-modal/eth-sign-modal.js]: Updated here https://github.com/MetaMask/metamask-extension/pull/18848/commits/00d7f01f4923628abea4c6eedbefbb9b1adad7ba

File: ui/components/app/modals/eth-sign-modal/eth-sign-modal.js [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+  const [showTextField, setShowTextField] = useState(false);
+  const [inputKeyword, setInputKeyword] = useState('');
+  const disabledRpcMethodPreferences = useSelector(
+    getDisabledRpcMethodPreferences,
+  );
+
+  const t = useI18nContext();
+  const dispatch = useDispatch();
+  const trackEvent = useContext(MetaMetricsContext);
+  const handleCancel = () => {
```
[the remark this answers is not in the document]
Comment (NidhiKJha) [reply ui/components/app/modals/eth-sign-modal/eth-sign-modal.js]: Thanks for pointing it out. Yeah, we can use hideModal directly Updated [here ](https://github.com/MetaMask/metamask-extension/pull/18848/commits/00d7f01f4923628abea4c6eedbefbb9b1adad7ba)

**Comment by NidhiKJha** [comment]
> Buttons look great now @NidhiKJha! only two more details: -Remove "try again" in the copy of the error state. -Make spacing between Icon and title 16 px and between title and body text 24px. 

@SaraCheikh  Updated! Need the builds to pass and the changes would be updated in the storybook then :)

**Comment by georgewrmarshall** [comment]
> So, the components/app/modals pattern and using dispatch to open modals should be deprecated and marked thusly. we do not want to extend this folder or its usages, at least the last time I checked. We had a popover component that can be used instead that doesn't require a dispatch to control showing it... unless that has changed recently? @georgewrmarshall what is the current approach recommended for modal implementation?

Hello @brad-decker @NidhiKJha, I was aware of both but never asked the question which we should use. 🤦 I was currently working on a Modal component that could be used for both so this is good information. May I suggest we add a deprecation notice to the modal.js folder so other engineers are aware? I'll create an issue https://github.com/MetaMask/metamask-extension/issues/18975

File: ui/components/app/modals/eth-sign-modal/index.scss [hunk 3]
```diff
@@ -0,0 +1,8 @@
+.eth-sign-modal {
+  position: relative;
+
+  &__close {
+    position: absolute;
+    right: 16px;
```
[the remark this answers is not in the document]
Comment (NidhiKJha) [reply ui/components/app/modals/eth-sign-modal/index.scss:6]: Yup, it works fine for RTL. This is the same logic we used for other modals

https://user-images.githubusercontent.com/39872794/235983169-3870956b-797c-45f5-acf1-20cb2acfff9f.mov

**Review by darkwing — APPROVED** [review]
This is *excellent* @NidhiKJha !  Well done!

**Review by pedronfigueiredo — APPROVED** [review]

**Pull request merged**